### PR TITLE
Adds citus_pause_node udf

### PR DESCRIPTION
**Pull request by gurkanindibay**
DESCRIPTION: Presenting citus_pause_node UDF enabling pausing by node_id.

citus_pause_node takes a node_id parameter and fetches all the shards in that node and puts AccessExclusiveLock on all the shards inside that node. With this lock, insert is disabled, until citus_pause_node transaction is closed.

### CODE REVIEW

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [inline src/backend/distributed/metadata/node_metadata.c:1345]: I have tests to cover this line and I can see that tests are being executed from the output. But still seeing below warning. What could be the reason?

Tests file result: https://github.com/citusdata/citus/blob/ed40dfe1a7482679aa4bc2068db19cba3655f337/src/test/regress/expected/isolation_citus_pause_node.out#L80C1-L111C18

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (aykut-bozkurt) [inline src/test/regress/spec/isolation_citus_pause_node.spec:170]: Can you add more permutations? e.g. 2 concurrent pause_node, pause_node-delete, pause_node-select (even if not blocked). You can also try a concurrent `start_metadata_sync_to_node`.

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (aykut-bozkurt) [inline src/test/regress/spec/isolation_citus_pause_node.spec:58]: You can use nextval(<sequence_name>) to get a non-existent node instead of the loop.

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (aykut-bozkurt) [inline src/test/regress/spec/isolation_citus_pause_node.spec:8]: Can you also add a reference table, Citus local table, and a tenant table?

File: src/backend/distributed/metadata/node_metadata.c
Comment (aykut-bozkurt) [inline src/backend/distributed/metadata/node_metadata.c:1339]: Should this method take nodename and nodeport as argument instead of nodeid for ease of use? Other similar udfs do so.

File: src/backend/distributed/metadata/node_metadata.c
Comment (aykut-bozkurt) [inline src/backend/distributed/metadata/node_metadata.c:1353]: This method would release all the locks at the end of transaction. I mean below execution would not pause the node at all.

```sql
select citus_pause_node(2);
```

But below would pause it until transaction ends.
```sql
begin;
select citus_pause_node(2);
```

I think the method should still pause even if we are not inside a transaction. You can wait for query cancellation inside a loop just after taking the locks. If user cancels, then return. What do you think?

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c:1339]: citus_update_node does not take node port and node_name. Since this UDF is like citus_update_node; working on an existing node; I think it is better work with id.

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c:1353]: citus_update_node with the begin block. I modeled that extension. @JelteF what do you think?

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c:1339]: Having both id and node name and port in parameters seems confusing to me BTW

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: added delete, select. Still trying to figure out start_metadata_sync_to_node

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: Fixed

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec:11]: Added a reference table. company is distributed table already. How can I add a local table into a worker node? In my case I need to add a local tavle into the worker node with id 3

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c:1353]: If we make such a change, we need to evaluate such UDFs as well, since it may be confusing for the users; some inside transaction, some not

Comment (onderkalaci) [reply src/backend/distributed/metadata/node_metadata.c:1353]: > citus_update_node is also used with the begin block

I think for citus_update_node it makes sense, because it is a one time operation, you update the metadata and you are done.

For citus_pause_node, that doesn't conceptually make sense. So, for me to understand more:

- what is the use case for `citus_update_node`? Is that mostly intended for use during failovers? If so, maybe the current approach might make sense, but we should enforce that the UDF is called within a transaction block
- Again, relevant to the above question. If this is intended for use during failovers, then it is fine that we block the transactions. However, if this is a generic UDFs that users can pause traffic on a node, acquiring locks would quickly block all backends, and net effect would be all traffic to the whole cluster is blocked (e.g., there are no free backends to continue querying the rest of the cluster, and/or we'd quickly hit max_connections)
- If we require transaction blocks, than some of the uses cases of citus_pause_node would not be supported. Say, you pause traffic to node 1, but then you also need to restart the coordinator. Then, the locks would be release

Comment (onderkalaci) [inline src/backend/distributed/metadata/node_metadata.c:1353]: Also, LockShardsInPlacementListMetadata seems to only acquire the locks on the coordinator, we should also acquire the same locks from the worker nodes as query from any node is the default behavior. 

We should also acquire the locks on the workers, you can check `BlockWritesToShardList` function for that

File: src/backend/distributed/metadata/node_metadata.c
Comment (onderkalaci) [inline src/backend/distributed/metadata/node_metadata.c:1350]: why only primary nodes? Can't we pause traffic to read-replicas? Is there anything that prevents us?

File: src/backend/distributed/metadata/node_metadata.c
Comment (JelteF) [reply src/backend/distributed/metadata/node_metadata.c:1339]: Yeah let's go for node_id. We've been switching to using node_id for some older UDFs too, since they are stable across failovers with Patroni.

File: src/backend/distributed/metadata/node_metadata.c
Comment (JelteF) [reply src/backend/distributed/metadata/node_metadata.c:1353]: > what is the use case for citus_update_node? Is that mostly intended for use during failovers?

(I'm going to assume you meant citus_pause_node here). It's initial intended use is indeed during failover. It's basically intended to do the locking of citus_update_node, but without having to specify a new hostname+port. The main usecase  here is for Patroni, because at the start of a switchover it does not know what node will be switched over to. But during a switchover we want to pause all the traffic to that worker so that no errors occur, only a little bit of latency increase.

So right now it pauses traffic using the quite hacky `citus_update_node(123, 'some-non-existing-hostname', 5432)`

Another usecase could be controlled restarts of  a worker. Where we pause traffic to the worker on the coordinator, while the restart is happening.

I do agree that the name is quite confusing. Maybe we should call it something like `citus_pause_node_during_transaction()` instead.

File: src/test/regress/isolation_schedule
Comment (JelteF) [inline src/test/regress/isolation_schedule:83]: unnecessary newline

File: src/backend/distributed/metadata/node_metadata.c
Comment (JelteF) [reply src/backend/distributed/metadata/node_metadata.c:1353]: I agree this is what we want, but I don't think calling BlockWritesToShardList would be enough. The main usecase of this `citus_pause_node` is calling it during failover, when `citus_update_node` is later used in the same transaction, and while one of the nodes is possibly down. A quick look at the code suggests that BlockWritesToShardList fails if it cannot block writes on one node. We wouldn't want that here. We'd want a best-effort locking on the workers, i.e. if one node is unreachable that's fine.

So while I agree we want this. It definitely feels out of scope for a warmup task. I think even just blocking on the coordinator (i.e. the current behaviour) has benefits enough for Patroni for this function to exist. The improvements of blocking on workers and blocking reads as well, seem like additions that we want. But seem better fit for follow up tasks.

Also, if we would do this in `citus_pause_node`, should we do the same thing in `citus_update_node` then?

Comment (onderkalaci) [reply src/backend/distributed/metadata/node_metadata.c:1353]: > I agree this is what we want, but I don't think calling BlockWritesToShardList would be enough.

Sure, I mostly shared to show that we need to acquire the same locks on the workers, not necessarily using this function.


> So while I agree we want this. It definitely feels out of scope for a warmup task

My feeling is that citus_pause_node might not be the best warm-up task anyway :) 

>  think even just blocking on the coordinator (i.e. the current behaviour) has benefits enough for Patroni for this function to exist. 

Well, my fear is that we'll keep building features that are not compatible with query from any node (which we started to do in Citus 12.0), then in a few years, we'd need re-iterate to support query-from-any node with all these known limitations.

Even for Patroni, if you are using query-from-any node now, this won't work well anyway.

> But seem better fit for follow up tasks.

Because of the above, it is probably not a follow-up task but part of the current task.


> Also, if we would do this in citus_pause_node, should we do the same thing in citus_update_node then?

yes, we should. 



Overall, I'm deferring the decisions to @JelteF and @gurkanindibay, but I think at a minimum we should pause the traffic from the workers as well.

Comment (onderkalaci) [reply src/backend/distributed/metadata/node_metadata.c:1353]: or citus_transaction_bound_pause_node()  / citus_pause_node_within_txn()

File: src/backend/distributed/metadata/node_metadata.c
Comment (JelteF) [inline src/backend/distributed/metadata/node_metadata.c:1350]: Let's extract the logic from citus_update_node and share it, this would also give us the possibility of adding `force`. It also allows us to later improve a single part of the code, to add support for pausing on worker nodes too.

File: src/backend/distributed/metadata/node_metadata.c
Comment (JelteF) [reply src/backend/distributed/metadata/node_metadata.c:1353]: > Well, my fear is that we'll keep building features that are not compatible with query from any node (which we started to do in Citus 12.0), then in a few years, we'd need re-iterate to support query-from-any node with all these known limitations.
> ...
> I think at a minimum we should pause the traffic from the workers as well.

tl;dr this is a new hook into already existing functionality, not a new feature. So, inheriting the bugs of the current functionality seems fine to me.

So, I don't think the difference is that I don't think of `citus_pause_node` as a new feature. Just as a new way of calling into the already existing node pausing functionality in citus_update_node. So I think it's fine if citus_pause_node inherits the bugs of citus_update_node it's pausing functionality (because indeed it's a bug that it doesn't pause on workers). To make that clearer, I suggested in another comment to reuse and share the actual pausing code of citus_update_node, instead of re-implementing it. 

Hopefully that helps with your worry that we'd need to do additional re-iteratation, because we currently already need to re-iterate the pausing logic of citus_update_node. So adding citus_pause_node now, wont impact the amount of re-iteration we need to do at all.

File: src/backend/distributed/metadata/node_metadata.c
Comment (JelteF) [reply src/backend/distributed/metadata/node_metadata.c:1350]: the original code in citus_update_node has a comment about this:
```
	 * It might be worth blocking reads to a secondary for the same reasons,
	 * though we currently only query secondaries on follower clusters
	 * where these locks will have no effect.
```

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c:1353]: Changing its name to  citus_pause_node_within_txn() @onderkalaci @JelteF is it ok?

Comment (JelteF) [reply src/backend/distributed/metadata/node_metadata.c:1353]: sounds good to me

Comment (onderkalaci) [reply src/backend/distributed/metadata/node_metadata.c:1353]: > . To make that clearer, I suggested in another comment to reuse and share the actual pausing code of citus_update_node, instead of re-implementing it.

sounds good, would be nice to add a comment saying that it is not compatible with query from any node

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: @aykut-bozkurt since we are locking the tables in pg_dist_placement, I don't understand why we are testing local tables. Do you want to see that local tables are not being locked?

Antoher one start_metadata_sync_to_node. I don't understand the necessity and scenario of testing this method? Could you clarify more?

File: src/test/regress/isolation_schedule
Comment (gurkanindibay) [reply src/test/regress/isolation_schedule]: Fixed

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c]: Is that comment for citus_pause_node_within_txn function

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (onderkalaci) [inline src/test/regress/spec/isolation_citus_pause_node.spec:232]: maybe add a test with SELECT on distributed table?

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (aykut-bozkurt) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: Yes, Postgres local tables are not in pg_dist_placement (metadata). I mean Citus local tables which are different than Postgres local tables in that we create a single placement at coordinator for them. They can be queried by other nodes (workers) as we store metadata for them. You can check `citus_add_local_table_to_metadata`.

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (aykut-bozkurt) [reply src/test/regress/spec/isolation_citus_pause_node.spec:11]: I gave info on Citus local tables above. You can try to pause coordinator after adding local table (its oly placement will be on coordinator). Then, a concurrent session should be blocked on any DML.

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (aykut-bozkurt) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: I wanted to see if metadata sync should be blocked or not. But I think we do not want to block it to finish the sync. We are only interested in blocking regular user queries. @JelteF Do you agree?

File: src/backend/distributed/metadata/node_metadata.c
Comment (onderkalaci) [inline src/backend/distributed/metadata/node_metadata.c:1167]: `Add lock` is an unfamiliar term. Maybe `Acquire shard metadata locks on all shards residing in the given worker node`

File: src/backend/distributed/metadata/node_metadata.c
Comment (onderkalaci) [inline src/backend/distributed/metadata/node_metadata.c:1346]: needs comment: What does this function do? Why within tx only?

We tend to forget these quickly, so better to document :)

File: src/backend/distributed/metadata/node_metadata.c
Comment (onderkalaci) [reply src/backend/distributed/metadata/node_metadata.c:1167]: also, this should definitely state something like the following (probably with better wording):
```
TODO: This function is not compatible with query from any node feature. Acquiring locks on placements should happen on all nodes, not only on the coordinator (or more precisely the node this function called)
```

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec:232]: Added

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c]: Added the comments

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c]: Added

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: @aykut-bozkurt citus_pause_node has a limited functionality to acquire locks. It acquires locks for all the shard placements for a specific node. Since local tables with citus metadata associated are defined only under coordinator, I don't think that these tables are in the scope of this task. @JelteF what do you think?

File: src/backend/distributed/sql/downgrades/citus--12.1-1--12.0-1.sql
Comment (hanefi) [inline src/backend/distributed/sql/downgrades/citus--12.1-1--12.0-1.sql:3]: I do not think that this statement is correct.

```suggestion
```

File: src/backend/distributed/sql/citus--12.0-1--12.1-1.sql
Comment (hanefi) [inline src/backend/distributed/sql/citus--12.0-1--12.1-1.sql:9]: How about adding the new line down below?

```suggestion
-- bump version to 12.1-1

#include "udfs/citus_pause_node_within_txn/12.1-1.sql"
#include "udfs/citus_prepare_pg_upgrade/12.1-1.sql"
```

**Review by aykut-bozkurt — APPROVED** [review]
LGTM. 
We can open an improvement issue to extend pausing to query from any node.

**Comment by onderkalaci** [comment]
> LGTM. We can open an improvement issue to extend pausing to query from any node.

@gurkanindibay please make sure to squash the commits before merging

**Comment by JelteF** [comment]
> please make sure to squash the commits before merging

Easiest way to do this is using "Squash and merge" from the dropdown next to the "Merge pull request" button

File: src/backend/distributed/metadata/node_metadata.c
Comment (gurkanindibay) [reply src/backend/distributed/metadata/node_metadata.c:1179]: Added the code you requested

File: src/backend/distributed/metadata/node_metadata.c
Comment (JelteF) [inline src/backend/distributed/metadata/node_metadata.c:1212]: this whole comment would fit better as a function comment on the actual function.

File: src/backend/distributed/metadata/node_metadata.c
Comment (JelteF) [inline src/backend/distributed/metadata/node_metadata.c:1185]: Function needs a comment

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (JelteF) [inline src/test/regress/spec/isolation_citus_pause_node.spec:266]: Let's show the difference between force and non-force using these tests:
```
-- s1-pause-node should block
"s1-begin" "s2-begin" "s2-insert-distributed" "s1-pause-node" "s2-end" "s1-end"

-- s1-pause-node-force
"s1-begin" "s2-begin" "s2-insert-distributed" "s1-pause-node-force"(*) "s1-end"  "s2-end"
```

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (JelteF) [inline src/test/regress/spec/isolation_citus_pause_node.spec:168]: This insert step can now contain only a single insert statement afaict (same for the insert-reference and delete-distributed step below).

```suggestion
	insert into employee values(11,'e11',3);
```

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (JelteF) [inline src/test/regress/spec/isolation_citus_pause_node.spec:102]: I don't think these notices are very helpful now. Let's remove them. 

```suggestion
	DO $$
	DECLARE
		v_shard_id int;
		v_node_id int;
		v_node_name text;
		v_node_port int;
	BEGIN
		-- Get the shard id for the distribution column
		SELECT get_shard_id_for_distribution_column('employee', 3) into v_shard_id;

		--Get the node id for the shard id
		SELECT nodename,nodeport into v_node_name,v_node_port FROM citus_shards WHERE shardid = v_shard_id limit 1;

		-- Get the node id for the shard id
		SELECT nodeid into v_node_id FROM pg_dist_node WHERE nodename = v_node_name and nodeport = v_node_port limit 1;


		-- Pause the node
		perform pg_catalog.citus_pause_node_within_txn(v_node_id) ;
	END;
```

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (JelteF) [inline src/test/regress/spec/isolation_citus_pause_node.spec:69]: What happens when you don't transform the ERROR to a NOTICE like this:
```suggestion
	BEGIN
		select nextval('pg_dist_node_nodeid_seq')::int into v_node_id;
		select citus_pause_node_within_txn(v_node_id) ;
```

I think we'll see the error in the test output, which would be a fine way to test, and reduced the testing code.

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (JelteF) [inline src/test/regress/spec/isolation_citus_pause_node.spec:219]: These should be swapped to test that the "force" argument kills the s2 query
```suggestion
permutation "s1-begin"  "s2-begin" "s2-insert-distributed" "s1-pause-node-force"(*) "s1-end" "s2-end"
```

Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec:185]:
```bash
test isolation_setup              ... ok           31 ms
test isolation_cluster_management ... ok           77 ms
test isolation_citus_pause_node   ... FAILED    11001 ms
```
In that case it takes 10 seconds to test the case. If we use so, I need to give a timeout; e.g. 100 ms

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec:69]: Node id is a changing value and error message includes node id. Therefore, I'm adding this exception handling to fix the error message

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: Fixed

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: Fixed

File: src/test/regress/spec/isolation_citus_pause_node.spec
Comment (gurkanindibay) [reply src/test/regress/spec/isolation_citus_pause_node.spec]: Used lock_cooldown parameter with the suggestion of @JelteF and it worked

**Review by JelteF — APPROVED** [review]

**Pull request merged**
